desc/equilibrium/coords.py

```diff
@@ -5,14 +5,143 @@
 import numpy as np
 from termcolor import colored
 
-from desc.backend import jit, jnp, put, while_loop
+from desc.backend import fori_loop, jit, jnp, put, while_loop
 from desc.compute import compute as compute_fun
-from desc.compute import get_transforms
-from desc.grid import ConcentricGrid, LinearGrid, QuadratureGrid
+from desc.compute import data_index, get_transforms
+from desc.grid import ConcentricGrid, Grid, LinearGrid, QuadratureGrid
 from desc.transform import Transform
 from desc.utils import Index
 
 
+def map_coordinates(eq, coords, inbasis, outbasis, tol=1e-6, maxiter=30, rhomin=1e-6):
+    """Given coordinates in inbasis, compute corresponding coordinates in outbasis.
+
+    First solves for the computational coordinates that correspond to inbasis, then
+    evaluates outbasis at those locations.
+
+    NOTE: this function cannot be JIT compiled or differentiated with AD.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium to use
+    coords : ndarray, shape(k,3)
+        2D array of input coordinates. Each row is a different
+        point in space.
+    inbasis, outbasis : tuple of str
+        Labels for input and output coordinates, eg ("R", "phi", "Z") or
+        ("rho", "alpha", "zeta") or any combination thereof. Labels should be the
+        same as the compute function data key
+    tol : float
+        Stopping tolerance.
+    maxiter : int > 0
+        Maximum number of Newton iterations
+    rhomin : float
+        Minimum allowable value of rho (to avoid singularity at rho=0)
+
+
+    """
+    assert (
+        np.isfinite(maxiter) and maxiter > 0
+    ), f"maxiter must be a positive integer, got {maxiter}"
+    assert np.isfinite(tol) and tol > 0, f"tol must be a positive float, got {tol}"
+
+    basis_derivs = [f"{X}_{d}" for X in inbasis for d in ("r", "t", "z")]
+    for key in basis_derivs:
+        assert (
+            key in data_index.keys()
+        ), f"don't have recipe to compute partial derivative {key}"
+
+    # can't use AD or jit compile this because of grid stuff
+    def compute(y, basis):
+        data = eq.compute(basis, grid=Grid(y, sort=False))
```

Comment:
Can you jit this by passing in the already constructed grid as parameter instead of y?

Reply:
in theory yes but the grid would need to be a static arg, so it would recompile on every iteration, so probably not worth it.